### PR TITLE
Support RevisionHistoryLimit for all ECK-managed resources

### LINKED ISSUE #5789: Support RevisionHistoryLimit in Deployment and Daemonset for Elastic Agents
## Proposal

As part of the `agents.agent.k8s.elastic.co` CRD under both deployment and daemonset RevisionHistoryLimit should be supported.

Deployment already supports `replicas`, `podTemplate` so adding a revisionHistoryLimit seems like a small lift.

**Use case. Why is this important?**

Being able to limit the number of `revisionHistoryLimit` matters for performance purposes, and cleaning up old replicasets, etc.

### PR DESCRIPTION
**Pull request by KnechtionsCoding**
Adds support for RevisionHistoryLimit to all ECK managed objects.

Similar to #5790 for all resources.
Closes #5789.

### CODE REVIEW

**Comment by thbkrkr** [comment]
Jenkins test this please

**Comment by thbkrkr** [comment]
I'm not sure it's really useful to have the `revisionHistoryLimit` field per NodeSet for Elasticsearch. Instead, we could have only one at the Spec level used for all StatefulSets?

File: pkg/apis/elasticsearch/v1/elasticsearch_types.go [hunk 1]
```diff
@@ -303,6 +303,9 @@ type NodeSet struct {
 	// +kubebuilder:pruning:PreserveUnknownFields
 	PodTemplate corev1.PodTemplateSpec `json:"podTemplate,omitempty"`
 
+	// RevisionHistoryLimit sets the number of old StatefulSets to retain to allow rollback.
```
Comment (pebrc) [inline pkg/apis/elasticsearch/v1/elasticsearch_types.go:306]:
```suggestion
	// RevisionHistoryLimit is the number of revisions to retain to allow rollback in the underlying StatefulSets.
```

File: pkg/apis/beat/v1beta1/beat_types.go [hunk 2]
```diff
@@ -94,6 +96,8 @@ type DeploymentSpec struct {
 	Replicas    *int32                 `json:"replicas,omitempty"`
 	// +kubebuilder:validation:Optional
 	Strategy appsv1.DeploymentStrategy `json:"strategy,omitempty"`
+	// RevisionHistoryLimit sets the number of old Deployment to retain to allow rollback.
```
Comment (pebrc) [inline pkg/apis/beat/v1beta1/beat_types.go:99]:
```suggestion
	// RevisionHistoryLimit is the number of revisions to retain to allow rollback in the underlying Deployment.
```

File: pkg/apis/beat/v1beta1/beat_types.go [hunk 3]
```diff
@@ -86,6 +86,8 @@ type DaemonSetSpec struct {
 	PodTemplate corev1.PodTemplateSpec `json:"podTemplate,omitempty"`
 	// +kubebuilder:validation:Optional
 	UpdateStrategy appsv1.DaemonSetUpdateStrategy `json:"updateStrategy,omitempty"`
+	// RevisionHistoryLimit sets the number of old DaemonSet to retain to allow rollback.
```
Comment (pebrc) [inline pkg/apis/beat/v1beta1/beat_types.go:89]:
```suggestion
	// RevisionHistoryLimit is the number of revisions to retain to allow rollback in the underlying DaemonSet.
```

File: pkg/apis/agent/v1alpha1/agent_types.go [hunk 4]
```diff
@@ -102,19 +102,19 @@ type Output struct {
...
 }
 
 type DeploymentSpec struct {
 	// +kubebuilder:pruning:PreserveUnknownFields
-	PodTemplate          corev1.PodTemplateSpec `json:"podTemplate,omitempty"`
-	Replicas             *int32                 `json:"replicas,omitempty"`
-	RevisionHistoryLimit *int32                 `json:"revisionHistoryLimit,omitempty"`
-
+	PodTemplate corev1.PodTemplateSpec `json:"podTemplate,omitempty"`
+	// RevisionHistoryLimit sets the number of old Deployment to retain to allow rollback.
```
Comment (pebrc) [inline pkg/apis/agent/v1alpha1/agent_types.go:115]:
```suggestion
	// RevisionHistoryLimit is the number of revisions to retain to allow rollback in the underlying Deployment.
```

File: pkg/apis/agent/v1alpha1/agent_types.go [hunk 5]
```diff
@@ -102,19 +102,19 @@ type Output struct {
 
 type DaemonSetSpec struct {
 	// +kubebuilder:pruning:PreserveUnknownFields
-	PodTemplate          corev1.PodTemplateSpec `json:"podTemplate,omitempty"`
-	RevisionHistoryLimit *int32                 `json:"revisionHistoryLimit,omitempty"`
-
+	PodTemplate corev1.PodTemplateSpec `json:"podTemplate,omitempty"`
+	// RevisionHistoryLimit sets the number of old DaemonSet to retain to allow rollback.
```
Comment (pebrc) [inline pkg/apis/agent/v1alpha1/agent_types.go:106]:
```suggestion
	// RevisionHistoryLimit is the number of revisions to retain to allow rollback in the underlying DaemonSet.
```

**Comment by thbkrkr** [comment]
Jenkins test this please

**Comment by thbkrkr** [comment]
Jenkins test this please

**Comment by KnechtionsCoding** [comment]
> I'm not sure it's really useful to have the `revisionHistoryLimit` field per NodeSet for Elasticsearch. Instead, we could have only one at the Spec level used for all StatefulSets?

I see two advantages to having it at the he nodeset level:

1. You could theoretically roll back master nodes separately from individual nodes.
2. It's where it is expected, i.e. alongside the podtemplate. That's where it is in the K8s spec. If someone wants to control them all together that's what the topology keys are for so you only need one node set per type?

But I'm not strongly attached to having it per node set so I'll move it up one level.

**Comment by KnechtionsCoding** [comment]
Yeah, because I wrote agent first and closed it in a separate PR. Would you like me to move back to the top level?

**Comment by thbkrkr** [comment]
@KnechtionsCoding Since I've already contributed to your PR and I'd like we close it, I'll take care of the last small change (move to the root level).

**Comment by thbkrkr** [comment]
Jenkins test this please

**Comment by KnechtionsCoding** [comment]
Oh, sure! I just was off this last week so couldn't take a look. Thanks for closing it out!

**Review by naemono — APPROVED** [review]
Tested all CRDs, and all configured the underlying components (sts, deployment, daemonset) properly, and became healthy.  👍

**Pull request merged**